### PR TITLE
Add updateFieldDescriptionMapping method to ModelManagerInterface

### PR DESCRIPTION
**Pull request by franmomu**
## Subject

Ref: https://github.com/sonata-project/SonataAdminBundle/pull/6249

There is an issue in the respective `ModelManager` of both persistence bundles (ORM and MongoDB), they both have `ModelManager::hasMetadata()` and `ModelManager::getMetadata()` methods that are called [from different places as they were in the interface](https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/search?q=hasMetadata&unscoped_q=hasMetadata).

Taking a closer look at these calls, in DoctrineMongoDBAdminBundle (ORM is equal or almost the same), they are doing something very similar:

**ShowBuilder**:
https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/blob/3a8e9d44b6e2c67436d2fd8d23aa5dd9b41a0e0a/src/Builder/ShowBuilder.php#L62-L75

**FormContractor**:
https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/blob/36290e0f1b4113a0e98084f1d4a977cbf476f54d/src/Builder/FormContractor.php#L51-L63

**ListBuilder**:
https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/blob/3a8e9d44b6e2c67436d2fd8d23aa5dd9b41a0e0a/src/Builder/ListBuilder.php#L68-L88

DatagridBuilder:
https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/blob/2be52edb46e449df710ae3d2e0b780ddeec020c3/src/Builder/DatagridBuilder.php#L68-L86

So the idea with this method is that it updates `fieldDecription` with the related mapping, so the implementation could look like (more or less):
```php
public function updateFieldDescriptionMapping(string $class, FieldDescriptionInterface $fieldDescription)
{
    if (!$this->hasMetadata($class)) {
        return;
    }

    [$metadata, $lastPropertyName, $parentAssociationMappings] = $admin->getModelManager()->getParentMetadataForProperty($class, $fieldDescription->getName());
    $fieldDescription->setParentAssociationMappings($parentAssociationMappings);

    // set the default field mapping
    if (isset($metadata->fieldMappings[$fieldDescription->getName()])) {
        $fieldDescription->setFieldMapping($metadata->fieldMappings[$lastPropertyName]);
    }

    // set the default association mapping
    if (isset($metadata->associationMappings[$fieldDescription->getName()])) {
        $fieldDescription->setAssociationMapping($metadata->associationMappings[$lastPropertyName]);
    }
}
``` 

There is another issue kind of related with `ModelManager::getParentMetadataForProperty()` [being also called as it was in the interface](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/fcd60598cc4fc6d385e2c900b60ed42622146ae0/src/Guesser/AbstractTypeGuesser.php#L31), but that would need more thought (could be done in another PR).

I am targeting this branch, because this change is BC.

Closes #6249.

## Changelog

```markdown
### Added
- Added `ModelManagerInterface::updateFieldDescriptionMapping()`.
```

### CODE REVIEW

File: src/Model/ModelManagerInterface.php [hunk 1]
```diff
@@ -338,4 +340,11 @@ public function getPaginationParameters(DatagridInterface $datagrid, $page);
      * @phpstan-param class-string $class
      */
     public function addIdentifiersToQuery($class, ProxyQueryInterface $query, array $idx);
+
+    // NEXT_MAJOR: Uncomment this method.
+//    /**
+//     * @phpstan-template T
```
Comment (jordisala1991) [inline src/Model/ModelManagerInterface.php:346]: No need to add template if there is no IN/OUT or more than one use of the T param.

File: src/Model/ModelManagerInterface.php [hunk 2]
```diff
@@ -338,4 +340,11 @@ public function getPaginationParameters(DatagridInterface $datagrid, $page);
...
+//     * @phpstan-param class-string<T> $class
+//     */
+//    public function updateFieldDescriptionMapping(string $class, FieldDescriptionInterface $fieldDescription): void;
```
Comment (VincentLanglet) [inline src/Model/ModelManagerInterface.php:349]: What about adding this method to a new FooInterface instead ?
Then ModelManagerInterface could extends FooInterface.

We could move some others methods, like
```
getNewFieldDescriptionInstance
```
which seems also `FieldDescription`-related

File: src/Model/ModelManagerInterface.php [hunk 3]
```diff
@@ -338,4 +340,11 @@ public function getPaginationParameters(DatagridInterface $datagrid, $page);
      * @phpstan-param class-string $class
      */
     public function addIdentifiersToQuery($class, ProxyQueryInterface $query, array $idx);
+
+    // NEXT_MAJOR: Uncomment this method.
+//    /**
```
Comment (VincentLanglet) [inline src/Model/ModelManagerInterface.php:345]: We should add a comment to explain the goal of this method.

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

**Comment by VincentLanglet** [comment]
> There is another issue kind of related with `ModelManager::getParentMetadataForProperty()` [being also called as it was in the interface](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/fcd60598cc4fc6d385e2c900b60ed42622146ae0/src/Guesser/AbstractTypeGuesser.php#L31), but that would need more thought (could be done in another PR).

IMHO it would be better if we find the solution now. 
Depending on the solution we find, this could radically change this PR.

**Comment by VincentLanglet** [comment]
> > There is another issue kind of related with `ModelManager::getParentMetadataForProperty()` [being also called as it was in the interface](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/fcd60598cc4fc6d385e2c900b60ed42622146ae0/src/Guesser/AbstractTypeGuesser.php#L31), but that would need more thought (could be done in another PR).
> 
> IMHO it would be better if we find the solution now.
> Depending on the solution we find, this could radically change this PR.

Any opinion/idea/advice on this, @sonata-project/contributors ?

**Comment by VincentLanglet** [comment]
Can we close this in favor of https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/pull/461 ?

**Comment by VincentLanglet** [comment]
Close in favor of https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/pull/492

**Pull request closed without merging**